packages/transaction-metrics-exporter/src/states.ts

```diff
@@ -0,0 +1,108 @@
+/* tslint:disable:ban-types */
+
+import { CeloContract, ContractKit } from '@celo/contractkit'
+
+type PromiseValue<T> = T extends PromiseLike<infer U> ? U : T
+type ContractWrapperType<C extends any> = PromiseValue<ReturnType<ContractKit['contracts'][C]>>
+type ArgumentTypes<F extends Function> = F extends (...args: infer A) => any ? A : never
+type ReturnMethodType<F extends Function> = F extends (...args: any) => infer A ? A : never
+
+export interface Contracts {
+  Exchange: ContractWrapperType<'getExchange'>
+  SortedOracles: ContractWrapperType<'getSortedOracles'>
+  Reserve: ContractWrapperType<'getReserve'>
+  GoldToken: ContractWrapperType<'getGoldToken'>
+  EpochRewards: ContractWrapperType<'getEpochRewards'>
+  StableToken: ContractWrapperType<'getStableToken'>
+}
+
+export interface StateGetter {
+  contract: keyof Contracts
+  method: string
+  args: any[] | ((blockNumber: number) => any[])
+  transformValues: Function
+  maxBucketSize: { [key: string]: number }
+}
+
+export function getter<
+  T extends keyof Contracts,
+  M extends keyof Contracts[T] & string,
+  RV extends string
+>(
+  contract: T,
+  method: M,
+  transformValues: (
+    state: PromiseValue<ReturnMethodType<Contracts[T][M] & Function>>
+  ) => { [key in RV]: any },
+  args:
+    | ArgumentTypes<Contracts[T][M] & Function>
+    | ((blockNumber: number) => ArgumentTypes<Contracts[T][M] & Function>) = [] as any,
+  maxBucketSize: { [key in RV]: any } = {} as any
+): StateGetter {
+  return { contract, method, args, transformValues, maxBucketSize }
+}
+
+export const stateGetters: StateGetter[] = [
```

Comment:
If I see correctly here, you are trying to generalize all state getters, but could you maybe elaborate on why you did so? As far as I can see, you are losing the Type benefits, the code becomes harder to read, and the metrics that are being logged become harder to reason about.

Reply:
- I did it to ensure that each metric is correct, uses the contract, method, arguments that are specified in the contract method
- I did this in this way because the previous code was too long and repetitive. With this getters, we have only the necessary information to add a metric, without repeating the same code several times.
- If the metrics that are logged are hard to understand, we can work on the result that is reported. Let me know how do you think that we should report it.

Reply:
What repetition are we exactly avoiding? Maybe I'm missing something, but actually just calling these methods to me does not sound that repetitive. I get that the motivation here was to get to a config-like state, however to a reader this is IMO more confusing since one has to jump back and forth to really grasp how this works.

Separately, we totally lose the compiler enforcement of the contractkit methods